### PR TITLE
chore: Remove explicit (bin0/bin1) access to MaterialSlab

### PR DESCRIPTION
**Pull request by asalzburger**
This PR removes the explicit:

`materialSlab(std::size_t bin0, std::size_t bin1)` access from the `ISurfaceMaterial` base class (and the derived classes), because it only makes sense for `BinnedSurfaceMaterial `, and not for other surface  material versions.

In all cases this was used, it was clear that you had `BinnedSurfaceMaterial` in hand, so you could directly access the `fullMaterial()` matrix.

### CODE REVIEW

**Comment by asalzburger** [comment]
Hi @ssdetlab - this is a tiny PR in preparation for the GridSurfaceMaterial, it removes an interface method that only makes sense for one specific implementation.

**Review by ssdetlab — APPROVED** [review]
Looks good to me. 

@asalzburger, I think that for the policy-bot to be satisfied I need to be on the reviewers list. I'll send the request

File: Core/src/Material/SurfaceMaterialMapper.cpp [hunk 1]
```diff
@@ -425,9 +425,14 @@ void Acts::SurfaceMaterialMapper::mapInteraction(
   for (auto tmapBin : touchedMapBins) {
     std::vector<std::array<std::size_t, 3>> trackBins = {tmapBin.second};
     if (m_cfg.computeVariance) {
-      tmapBin.first->trackVariance(
-          trackBins, touchedMaterialBin[tmapBin.first]->materialSlab(
-                         trackBins[0][0], trackBins[0][1]));
+      // This only makes sense for the binned material
+      auto binnedMaterial = dynamic_cast<const BinnedSurfaceMaterial*>(
+          touchedMaterialBin[tmapBin.first].get());
```
Comment (andiwand) [inline Core/src/Material/SurfaceMaterialMapper.cpp:430]: would it make sense to distinguish between non-`BinnedSurfaceMaterial` and `nullptr`?

File: Core/src/Material/SurfaceMaterialMapper.cpp [hunk 2]
```diff
@@ -498,11 +503,15 @@ void Acts::SurfaceMaterialMapper::mapSurfaceInteraction(
...
     std::vector<std::array<std::size_t, 3>> trackBins = {tmapBin.second};
     if (m_cfg.computeVariance) {
-      tmapBin.first->trackVariance(
-          trackBins,
-          touchedMaterialBin[tmapBin.first]->materialSlab(trackBins[0][0],
-                                                          trackBins[0][1]),
-          true);
+      // This only makes sense for the binned material
+      auto binnedMaterial = dynamic_cast<const BinnedSurfaceMaterial*>(
+          touchedMaterialBin[tmapBin.first].get());
```
Comment (andiwand) [inline Core/src/Material/SurfaceMaterialMapper.cpp:508]: same here

File: Examples/Io/Root/src/RootMaterialWriter.cpp [hunk 3]
```diff
@@ -156,11 +158,10 @@ void ActsExamples::RootMaterialWriter::writeMaterial(
                          -0.5, bins0 - 0.5, bins1, -0.5, bins1 - 0.5);
 
     // loop over the material and fill
-    for (std::size_t b0 = 0; b0 < bins0; ++b0) {
-      for (std::size_t b1 = 0; b1 < bins1; ++b1) {
-        // get the material for the bin
-        auto& mat = sMaterial->materialSlab(b0, b1);
-        if (mat) {
+    if (bsm != nullptr) {
```
Comment (andiwand) [inline Examples/Io/Root/src/RootMaterialWriter.cpp:161]: after line 99 this should always evaluate to `true` no?

File: Examples/Io/Root/src/RootMaterialWriter.cpp [hunk 4]
```diff
@@ -169,6 +170,15 @@ void ActsExamples::RootMaterialWriter::writeMaterial(
...
       }
+    } else if (bins1 == 1 && bins0 == 1) {
+      // homogeneous surface
+      auto mat = sMaterial->materialSlab(Acts::Vector3{0, 0, 0});
+      t->SetBinContent(1, 1, mat.thickness());
+      x0->SetBinContent(1, 1, mat.material().X0());
+      l0->SetBinContent(1, 1, mat.material().L0());
+      A->SetBinContent(1, 1, mat.material().Ar());
+      Z->SetBinContent(1, 1, mat.material().Z());
+      rho->SetBinContent(1, 1, mat.material().massDensity());
```
Comment (andiwand) [inline Examples/Io/Root/src/RootMaterialWriter.cpp:181]: just for me to understand: what is that special treatment for?

File: Core/src/Material/SurfaceMaterialMapper.cpp (hunk 1 above)
Comment (asalzburger) [reply Core/src/Material/SurfaceMaterialMapper.cpp:430]: The following PRs (from the MM) series, will actually clean this up.

I.e. you the `SurfaceMaterialMapper` will turn into `BinnedSurfaceMaterialAcummulator` - so the type will be clear here.

File: Examples/Io/Root/src/RootMaterialWriter.cpp (hunk 4 above)
Comment (asalzburger) [reply Examples/Io/Root/src/RootMaterialWriter.cpp:181]: We are writing also the HomogeneousMaterial to a 1-1 binned root structure, in order to have identical trees.

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [4cb1e5b15f0f1a3b912541649e69ebb1694ae3c1]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098754)||
|:green_circle: | [test_ActsEFTrackFit](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098753)||
|:green_circle: | [test_ActsBenchmarkWithSpot](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098752)||
|:green_circle: | [test_ActsConversionWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098751)||
|:green_circle: | [test_ActsWorkflowFastTracking](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098750)||
|:green_circle: | [test_ActsWorkflow](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098749)||
|:green_circle: | [test_ActsValidateAmbiguityResolution](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098748)||
|:green_circle: | [test_ActsValidateResolvedTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098747)||
|:green_circle: | [test_ActsValidateTracks](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098746)||
|:green_circle: | [test_ActsValidateActsCoreSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098745)||
|:green_circle: | [test_ActsValidateActsSpacePoints](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098744)||
|:green_circle: | [test_ActsValidateSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098743)||
|:green_circle: | [test_ActsValidateOrthogonalSeeds](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098742)||
|:green_circle: | [test_ActsValidateClusters](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098741)||
|:green_circle: | [test_ActsPersistifyEDM](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098740)||
|:green_circle: | [test_ActsGSFRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098739)||
|:green_circle: | [test_ActsKfRefitting](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098738)||
|:green_circle: | [test_ActsExtrapolationAlgTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098737)||
|:green_circle: | [test_ActsITkTest](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098736)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098735)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098734)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098733)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098732)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098731)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098730)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/37098729)||

**Pull request merged**